### PR TITLE
CINE: Add support for Operation Stealth.

### PR DESCRIPTION
**Pull request by karisal**
Hello!
Here's support for Operation Stealth PC 16 and 256 color versions with AdLib and Roland MT-32 sound.
The support is based on disassembly of PC versions of Future Wars and Operation Stealth using Interactive Disassembler.
Unfortunately the support is just a single commit because I simply worked on it locally until it worked. I'm just now starting to learn using Git (Previously I used SVN) but already it looks like it's a lot more powerful than SVN.

I think with this Operation Stealth is ready for testing, at least the PC versions. Amiga versions may work (Not sure about their audio side). Atari ST versions are completely untested.

Kari Salminen

### CODE REVIEW

**Comment by karisal** [comment]
Now I'm just waiting for somebody (e.g. @sev- or anybody) to review or merge this. Thank you very much!

Kari Salminen

File: engines/cine/anim.cpp [hunk 1]
```diff
@@ -798,8 +816,34 @@ int loadSeq(const char *resourceName, int16 idx) {
  */
 int loadResource(const char *resourceName, int16 idx, int16 frameIndex) {
 	int result = -1; // Return an error by default
+
+	if (g_cine->getGameType() == Cine::GType_OS &&
+		g_cine->getPlatform() == Common::kPlatformDOS &&
+		g_sound->musicType() != MT_MT32 &&
+		(strstr(resourceName, ".SPL") || strstr(resourceName, ".H32")))
+	{
```
Comment (sev-) [inline engines/cine/anim.cpp:824]: The code formatting is off. The parentheses must stick to the previous line.

File: engines/cine/anim.cpp [hunk 2]
```diff
@@ -798,8 +816,34 @@ int loadSeq(const char *resourceName, int16 idx) {
...
+		char base[20];
+		removeExtention(base, resourceName);
+
+		for (int i = 0; i < ARRAYSIZE(resNameMapping); i++) {
+			if (scumm_stricmp(base, resNameMapping[i].from) == 0) {
+				strncpy(base, resNameMapping[i].to, sizeof(base));
```
Comment (sev-) [inline engines/cine/anim.cpp:830]: This is incorrect. Either use `sizeof(base) - 1` or better use `Common::strlcpy()`

File: engines/cine/anim.cpp [hunk 3]
```diff
@@ -798,8 +816,34 @@ int loadSeq(const char *resourceName, int16 idx) {
...
+				break;
+			}
+		}
+
+		const char* ext = (g_sound->musicType() == MT_ADLIB) ? ".ADL" : ".HP";
```
Comment (sev-) [inline engines/cine/anim.cpp:835]: Incorrect code formatting. `char *ext`

File: engines/cine/anim.cpp [hunk 4]
```diff
@@ -798,8 +816,34 @@ int loadSeq(const char *resourceName, int16 idx) {
...
+
+	if (g_cine->getGameType() == Cine::GType_OS &&
+		g_cine->getPlatform() == Common::kPlatformDOS &&
+		g_sound->musicType() != MT_MT32 &&
+		(strstr(resourceName, ".SPL") || strstr(resourceName, ".H32")))
+	{
+		char base[20];
+		removeExtention(base, resourceName);
+
+		for (int i = 0; i < ARRAYSIZE(resNameMapping); i++) {
```
Comment (sev-) [inline engines/cine/anim.cpp:828]: This should be `uint`

File: engines/cine/detection.cpp [hunk 5]
```diff
@@ -197,12 +327,11 @@ void CineMetaEngine::removeSaveState(const char *target, int slot) const {
...
 	}
 
-	out->write(saveNames, 10 * 20);
+	out->write(saveNames, SAVELIST_SIZE);
 	delete out;
 
 	// Delete save file
-	char saveFileName[256];
-	sprintf(saveFileName, "%s.%1d", target, slot);
+	const char * saveFileName = getSavegameFile(slot, target);
```
Comment (sev-) [inline engines/cine/detection.cpp:334]: `const char *saveFileName`

File: engines/cine/main_loop.cpp [hunk 6]
```diff
@@ -166,24 +175,39 @@ static void processEvent(Common::Event &event) {
 		case Common::KEYCODE_KP3:
 			moveUsingKeyboard(+1, -1); // Down & Right
 			break;
+		case Common::KEYCODE_LEFT:
```
Comment (sev-) [inline engines/cine/main_loop.cpp:178]: Maybe it is too much to ask, but if you could take a look into the new keymapper code (see Wintermute engine), that would be awesome.

File: engines/cine/pal.cpp [hunk 7]
```diff
@@ -140,19 +140,29 @@ int power(int base, int power) {
...
+
+	return *this;
+}
+
+Palette &Palette::rotateLeft(byte firstIndex, byte lastIndex) {
+	debug(1, "Palette::rotateLeft(firstIndex: %d, lastIndex: %d)", firstIndex, lastIndex);
+
+	const Color firstColor = _colors[firstIndex];
+
+	for (int i = firstIndex; i < lastIndex; i++)
```
Comment (sev-) [inline engines/cine/pal.cpp:161]: This should be `uint`

File: engines/cine/main_loop.cpp (hunk 6 above)
Comment (sev-) [reply engines/cine/main_loop.cpp:178]: Ah, here is how to do it: https://wiki.scummvm.org/index.php?title=Keymapper

File: engines/cine/pal.cpp (hunk 7 above)
Comment (karisal) [reply engines/cine/pal.cpp]: Taken into account.

File: engines/cine/detection.cpp (hunk 5 above)
Comment (karisal) [reply engines/cine/detection.cpp]: Taken into account.

File: engines/cine/anim.cpp (hunk 4 above)
Comment (karisal) [reply engines/cine/anim.cpp]: Taken into account.

File: engines/cine/anim.cpp (hunk 3 above)
Comment (karisal) [reply engines/cine/anim.cpp]: Taken into account.

File: engines/cine/anim.cpp (hunk 2 above)
Comment (karisal) [reply engines/cine/anim.cpp]: Taken into account (strlcpy).

File: engines/cine/anim.cpp (hunk 1 above)
Comment (karisal) [reply engines/cine/anim.cpp]: Taken into account.

File: engines/cine/main_loop.cpp (hunk 6 above)
Comment (karisal) [reply engines/cine/main_loop.cpp:178]: Yes, this is too much to ask at this point. So, not taken into account.

**Comment by karisal** [comment]
> Incredible comeback, Kari! Welcome!

  Thanks!

> There are a few minor issues I spotted and it will be ready for the merge.

  I submitted a commit fixing all but the keymapper code.

> Will you hang around for announcing support and fixing any Coverity (static analysis tool) issues if any?

  I can fix the Coverity issues if any come up recently after merge.
I can be around when support is announced (I don't need to be the
one who announces support, it can be someone else). I can perhaps
(Time permitting) be around for the bugfixing/testing period of the
first release to include this code but after that I will be gone i.e. not
available for any work related to ScummVM.

Kari Salminen

**Comment by sev-** [comment]
Great! I am merging this and will announce the support in couple of days.

**Comment by sev-** [comment]
@karisal Could you please make 3-4 interesting screenshots which I could put to the announcement on Facebook? One could be the game title.

**Comment by karisal** [comment]
Hi @sev,
sorry, I am busy for the next 8 days
at least and can't help. Maybe someone
else can do this in the meanwhile or
if it is not urgent then I might be able
to look into this later. My schedule is
quite unpredictable and/or busy
right now for the next couple of
weeks.

su 9. elok. 2020 klo 0.24 Eugene Sandulenko <notifications@github.com>
kirjoitti:

**Comment by karisal** [comment]
> 
> 
> @karisal Could you please make 3-4 interesting screenshots which I could put to the announcement on Facebook? One could be the game title.

Voila! If you didn't already make an announcement here are some screenshots (OptiPNG + AdvPNG processed):

[scummvm-os-screenshots.zip](https://github.com/scummvm/scummvm/files/5093104/scummvm-os-screenshots.zip)

**Pull request merged**